File: Sources/ConnectorControlCore/Slug.swift
/// File names for collection documents: lowercase ASCII letters and digits, every other run
/// collapsed to one hyphen, so a Mac, a PC and a case-sensitive git server agree on the name.
public enum Slug {
    public static func make(_ name: String) -> String {
        var out = ""
        var pendingHyphen = false
        // Lowercased as a whole, so a letter whose lowercase is longer expands: "İ" becomes "i" and a
        // combining dot, which slugs "İstanbul" as "i-stanbul" where Windows, whose invariant
        // lowercase leaves "İ" as it is, writes "stanbul". Harmless: the slug is fixed once, in the
        // sidecar, by whichever machine publishes first, and never derived again.
        for scalar in name.lowercased().unicodeScalars {
            let v = scalar.value
            let isAlnum = (v >= 0x61 && v <= 0x7A) || (v >= 0x30 && v <= 0x39)
            if isAlnum {
                if pendingHyphen, !out.isEmpty { out.append("-") }
                pendingHyphen = false
                out.unicodeScalars.append(scalar)
            } else {
                pendingHyphen = true
            }
        }
        return out.isEmpty ? "collection" : out
    }
}

File: windows/src/ConnectorControl.Core/Slug.cs
using System.Text;

namespace ConnectorControl.Core;

/// <summary>
/// File names for collection documents: lowercase ASCII letters and digits, every other run
/// collapsed to one hyphen, so a Mac, a PC and a case-sensitive git server agree on the name.
/// </summary>
public static class Slug
{
    public static string Make(string name)
    {
        var sb = new StringBuilder();
        var pendingHyphen = false;
        // The invariant lowercase leaves "İ" as it is, so "İstanbul" slugs as "stanbul", where the
        // Mac, lowering the whole string, expands it to "i" and a combining dot and writes
        // "i-stanbul". Harmless: the slug is fixed once, in the sidecar, by whichever machine
        // publishes first, and never derived again.
        foreach (var ch in name.ToLowerInvariant())
        {
            var isAlnum = (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9');
            if (isAlnum)
            {
                if (pendingHyphen && sb.Length > 0)
                {
                    sb.Append('-');
                }
                pendingHyphen = false;
                sb.Append(ch);
            }
            else
            {
                pendingHyphen = true;
            }
        }
        return sb.Length == 0 ? "collection" : sb.ToString();
    }
}
